src/charm.py

```diff
@@ -117,98 +94,123 @@ def __init__(self, *args):
+        # start kafka service
+        self.workload.start()
+        logger.info("Kafka snap started")
 
-        Returns:
-            True if ZK is related and `sync` user has been added. False otherwise.
-        """
-        if not self.peer_relation:
-            self._set_status(Status.NO_PEER_RELATION)
-            return False
+        # check for connection
+        self._on_update_status(event)
 
-        if not self.kafka_config.zookeeper_related:
-            self._set_status(Status.ZK_NOT_RELATED)
-            return False
+        # only log once on successful 'on-start' run
+        if isinstance(self.unit.status, ActiveStatus):
+            logger.info(f'Broker {self.unit.name.split("/")[1]} connected')
 
-        if not self.kafka_config.zookeeper_connected:
-            self._set_status(Status.ZK_NO_DATA)
-            return False
+    def _on_config_changed(self, event: EventBase) -> None:
+        """Generic handler for most `config_changed` events across relations."""
+        # only overwrite properties if service is already active
+        if not self.healthy or not self.upgrade.idle:
+            event.defer()
+            return
 
-        # TLS must be enabled for Kafka and ZK or disabled for both
-        if self.tls.enabled ^ (
-            self.kafka_config.zookeeper_config.get("tls", "disabled") == "enabled"
-        ):
-            self._set_status(Status.ZK_TLS_MISMATCH)
-            return False
+        # Load current properties set in the charm workload
+        properties = self.workload.read(self.workload.paths.server_properties)
+        properties_changed = set(properties) ^ set(self.config_manager.server_properties)
 
-        if not self.kafka_config.internal_user_credentials:
-            self._set_status(Status.NO_BROKER_CREDS)
-            return False
+        zk_jaas = self.workload.read(self.workload.paths.zk_jaas)
+        zk_jaas_changed = set(zk_jaas) ^ set(self.config_manager.zk_jaas_config.splitlines())
 
-        return True
+        if not properties and zk_jaas:
+            # Event fired before charm has properly started
+            event.defer()
+            return
 
-    @property
-    def healthy(self) -> bool:
-        """Checks and updates various charm lifecycle states.
+        # update environment
+        self.config_manager.set_environment()
 
-        Is slow to fail due to retries, to be used sparingly.
+        if zk_jaas_changed:
+            clean_broker_jaas = [conf.strip() for conf in zk_jaas]
+            clean_config_jaas = [
+                conf.strip() for conf in self.config_manager.zk_jaas_config.splitlines()
+            ]
+            logger.info(
+                (
+                    f'Broker {self.unit.name.split("/")[1]} updating JAAS config - '
+                    f"OLD JAAS = {set(clean_broker_jaas) - set(clean_config_jaas)}, "
+                    f"NEW JAAS = {set(clean_config_jaas) - set(clean_broker_jaas)}"
```

Comment:
**todo:** This was my fault as it was my change, but I think this might make Kafka restart endlessly. 

We can fix with:

```python
clean_broker_jaas = [conf.strip() for conf in self.workload.read(self.workload.paths.zk_jaas)]
clean_config_jaas = [
    conf.strip() for conf in self.config_manager.zk_jaas_config.splitlines()
]

zk_jaas_changed = set(clean_broker_jaas) ^ set(clean_config_jaas)

...

if zk_jaas_changed:
   logger.info(...)
```